operator/pkg/reconciliation/constructor.go

```diff
@@ -367,6 +368,9 @@ func setOperatorProgressStatus(rc *ReconciliationContext, newState api.ProgressS
 
 	patch := client.MergeFrom(rc.Datacenter.DeepCopy())
 	rc.Datacenter.Status.CassandraOperatorProgress = newState
+	if newState == api.ProgressReady {
+		rc.Datacenter.Status.ObservedGeneration = rc.Datacenter.Generation
```

Comment:
💯

Reply:
Err, actually, I don't think this is quite right. The ObservedGeneration and the actual generation being equal doesn't mean the resource is ready. All it means is that the operator has at least seen the generation in question. For example, when the ObservedGeneration for a statefulset equals the actual generation, that does not mean the statefulset is up and running, it just means the controller has had a chance to respond to the change of that resource.

Reply:
right - we've decided it's ready and then we set the ObservedGeneration - this is to provide something better than what we have right now, where you don't know if the CassDC is ready because the operator hasn't started work on changing it, or if it's all done in under a couple milliseconds

Someone working on automation around Cass Operator chatted me the other day...

> Matching generation would mean like "the requested spec has been successfully fully propagated at least once"

Reply:
I think the readiness of the cluster and the value of the observed generation are orthogonal. If you want to know if the requested spec has been fully propagated, you would check the `DatacenterReady` condition _and_ check the observed generation. The status tells you what the current state is, the observed generation tells you if the status is up-to-date. No one should rely exclusively on the observed generation to know if the current spec is fully propagated. That's not how statefulsets work for example. The observed generation equalling the current generation doesn't really have anything to do with readiness, it just means the operator has seen the current generation.

Reply:
That's right - check both - sorry I didn't get that exact phrase out in my last reply 😇

Reply:
I guess I'm still confused as to why we are only doing this when `ProgressReady`. We should do this regardless of the state at the end of reconciliation (unless we are in a quiet period). The value of `CassandraOperatorProgress` does not seem relevant here.

Reply:
We haven't fully observed and pushed down the change until all our writes to k8s resources are done, right?